homeassistant/components/light/template.py

```diff
@@ -0,0 +1,192 @@
+            entity_ids = (device_config.get(ATTR_ENTITY_ID) or
+                          state_template.extract_entities())
+            state_template.hass = hass
+
+        lights.append(
+            LightTemplate(
+                hass, device, friendly_name, state_template,
+                on_action, off_action, level_action, level_template,
+                entity_ids)
+        )
+
+    if not lights:
+        _LOGGER.error("No lights added")
+        return False
+
+    async_add_devices(lights, True)
+    return True
+
+
+class LightTemplate(Light):
+    """Representation of a templated Light, including dimmable."""
+
+    def __init__(self, hass, device_id, friendly_name, state_template,
+                 on_action, off_action, level_action, level_template,
+                 entity_ids):
+        """Initialize the light."""
+        self.hass = hass
+        self.entity_id = async_generate_entity_id(
+            ENTITY_ID_FORMAT, device_id, hass=hass)
+        self._name = friendly_name
+        self._template = state_template
+        self._on_script = Script(hass, on_action)
+        self._off_script = Script(hass, off_action)
+        self._level_script = Script(hass, level_action)
+        self._level_template = level_template
+
+        self._state = False
+        self._brightness = None
+        self._entities = entity_ids
+
+        if self._template is not None:
+            self._template.hass = self.hass
+        if self._level_template is not None:
+            self._level_template.hass = self.hass
+
+    @property
+    def brightness(self):
+        """Return the brightness of the light."""
+        return self._brightness
+
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        if self._level_script is not None:
+            return SUPPORT_BRIGHTNESS
+
+        return 0
+
+    @property
+    def is_on(self):
+        """Return true if device is on."""
+        return self._state
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Register callbacks."""
+        state = yield from async_get_last_state(self.hass, self.entity_id)
+        if state:
+            self._state = state.state == STATE_ON
+
+        @callback
+        def template_light_state_listener(entity, old_state, new_state):
+            """Handle target device state changes."""
+            self.hass.async_add_job(self.async_update_ha_state(True))
+
+        @callback
+        def template_light_startup(event):
+            """Update template on startup."""
+            async_track_state_change(
+                self.hass, self._entities, template_light_state_listener)
+
+            self.hass.async_add_job(self.async_update_ha_state(True))
+
+        self.hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_START, template_light_startup)
+
+    @asyncio.coroutine
+    def async_turn_on(self, **kwargs):
+        """Turn the light on."""
+        self._state = True
```

Comment:
Only optimistically set the status if there is no template defined. Same for brightness and turn_off below. If the status is optimistically updated, run `self.hass.async_add_job(self.async_update_ha_state())` to update the state. Note that we don't need to pass `True`, since `async_update` isn't necessary.